orbit/pkg/useraction/mdm_migration_darwin.go

```diff
@@ -0,0 +1,287 @@
+// without blocking.
+//
+// The first returned channel sends the exit code returned by swiftDialog, and
+// the second channel is used to send errors.
+func (b *baseDialog) render(flags ...string) (chan swiftDialogExitCode, chan error) {
+	exitCodeChan := make(chan swiftDialogExitCode)
+	errChan := make(chan error)
+	go func() {
+		cmd := exec.Command(b.path, flags...) //nolint:gosec
+		done := make(chan error)
+
+		if err := cmd.Start(); err != nil {
+			errChan <- err
+			return
+		}
+
+		go func() { done <- cmd.Wait() }()
+
+		select {
+		case <-b.interruptCh:
+			if err := cmd.Process.Signal(os.Interrupt); err != nil {
+				log.Error().Err(err).Msg("sending interrupt signal to swiftDialog process")
+				if err := cmd.Process.Kill(); err != nil {
+					log.Error().Err(err).Msg("killing swiftDialog process")
+				}
+
+			}
+		case err := <-done:
+			if err != nil {
+				// non-zero exit codes
+				if exitError, ok := err.(*exec.ExitError); ok {
+					ec := exitError.ExitCode()
+					switch ec {
+					case errorExitCode:
+						exitCodeChan <- errorExitCode
+					case secondaryBtnExitCode, infoBtnExitCode, timeoutExitCode:
+						exitCodeChan <- swiftDialogExitCode(ec)
+					default:
+						errChan <- fmt.Errorf("unknown exit code showing dialog: %w", exitError)
+					}
+				}
+
+				errChan <- fmt.Errorf("running swiftDialog: %w", err)
+			}
+
+			exitCodeChan <- 0
+		}
+
+	}()
+	return exitCodeChan, errChan
+}
+
+func NewMDMMigrator(path string, frequency time.Duration, handler MDMMigratorHandler) MDMMigrator {
+	return &swiftDialogMDMMigrator{
+		handler:    handler,
+		baseDialog: newBaseDialog(path),
+		frequency:  frequency,
+	}
+}
+
+// swiftDialogMDMMigrator implements MDMMigrator for macOS using swiftDialog as
+// the underlying mechanism for user action.
+type swiftDialogMDMMigrator struct {
+	*baseDialog
+	props     MDMMigratorProps
+	frequency time.Duration
+	handler   MDMMigratorHandler
+
+	// ensures only one dialog is open at a time, protects access to
+	// lastShown
+	showMu    sync.Mutex
+	lastShown time.Time
+
+	// ensures only one dialog is open at a given interval
+	intervalMu sync.Mutex
+}
+
+func (m *swiftDialogMDMMigrator) render(message string, flags ...string) (chan swiftDialogExitCode, chan error) {
+	icon := m.props.OrgInfo.OrgLogoURL
+	if icon == "" {
+		icon = "http://localhost:64793/fleet-mark-color-40x40@4x.png"
+	}
+
+	flags = append([]string{
+		// disable the built-in title so we have full control over the
+		// content
+		"--title", "none",
+		// top icon
+		"--icon", icon,
+		"--iconsize", "80",
+		"--centreicon",
+		// modal content
+		"--message", message,
+		"--messagefont", "size=16",
+		"--alignment", "center",
+		"--ontop",
+	}, flags...)
+
+	return m.baseDialog.render(flags...)
+}
+
+func (m *swiftDialogMDMMigrator) renderLoadingSpinner() (chan swiftDialogExitCode, chan error) {
+	return m.render("## Migrate to Fleet\n\nCommunicating with MDM server...",
```

Comment:
This gets called from `renderMigration` but it never receives from the channels, meaning that the render goroutine would block forever and leak. Using buffered channels of 1 would fix that.